buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java

```diff
@@ -105,87 +117,195 @@ private static void setupRootJdkDownload(Project rootProject, String platform, S
+            Action<CopySpec> removeRootDir = copy -> {
+                // remove extra unnecessary directory levels
+                copy.eachFile(details -> {
+                    /*
+                     * We want to remove up to the and including the jdk-.* relative paths. That is a JDK archive is structured as:
+                     *   jdk-12.0.1/
+                     *   jdk-12.0.1/Contents
+                     *   ...
+                     *
+                     * and we want to remove the leading jdk-12.0.1. Note however that there could also be a leading ./ as in
+                     *   ./
+                     *   ./jdk-12.0.1/
+                     *   ./jdk-12.0.1/Contents
+                     *
+                     * so we account for this and search the path components until we find the jdk-12.0.1, and strip the leading components.
+                     */
+                    String[] pathSegments = details.getRelativePath().getSegments();
+                    int index = 0;
+                    for (; index < pathSegments.length; index++) {
+                        if (pathSegments[index].matches("jdk-.*")) break;
+                    }
+                    assert index + 1 <= pathSegments.length;
+                    String[] newPathSegments = Arrays.copyOfRange(pathSegments, index + 1, pathSegments.length);
+                    details.setRelativePath(new RelativePath(true, newPathSegments));
+                });
+                copy.setIncludeEmptyDirs(false);
+            };
+            extractTask = rootProject.getTasks().register(extractTaskName, Copy.class, copyTask -> {
+                copyTask.doFirst(new Action<Task>() {
+                    @Override
+                    public void execute(Task t) {
+                        rootProject.delete(extractDir);
+                    }
+                });
+                copyTask.into(extractDir);
+                copyTask.from(fileGetter, removeRootDir);
+            });
         } else {
-            fileGetter = () -> rootProject.tarTree(rootProject.getResources().gzip(jdkArchiveGetter.get()));
-        }
-        String extractDir = rootProject.getBuildDir().toPath().resolve("jdks/openjdk-" + jdkVersion + "_" + platform).toString();
-        TaskProvider<Copy> extractTask = rootProject.getTasks().register(extractTaskName, Copy.class, copyTask -> {
-            copyTask.doFirst(new Action<Task>() {
-                @Override
-                public void execute(Task t) {
-                    rootProject.delete(extractDir);
-                }
+            /*
+             * Gradle TarFileTree does not resolve symlinks, so we have to manually extract and preserve the symlinks.
+             * cf. https://github.com/gradle/gradle/issues/3982 and https://discuss.gradle.org/t/tar-and-untar-losing-symbolic-links/2039
+             */
+            extractTask = rootProject.getTasks().register(extractTaskName, Task.class, task -> {
```

Comment:
Can we refactor this into a `SymlinkPreservingUntarTask` and map proper inputs/outputs so this can be incremental?

Reply:
I pushed 813a9bde0bffb280a600c2d5b722186369aa5e58.